ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java

```diff
@@ -197,4 +225,313 @@ public void updateAccounts(Collection<Account> accounts, int batchSize) throws S
+
+        for (Container container : updatedContainers) {
+          updateContainerStatement.setString(1, container.toJson().toString());
+          updateContainerStatement.setInt(2, container.getSnapshotVersion());
+          updateContainerStatement.setInt(3, account.getId());
+          updateContainerStatement.setInt(4, container.getId());
+          updateContainerStatement.addBatch();
+          ++updateContainerCount;
+          if (++batchCount >= batchSize) {
+            executeBatchUpdates(insertAccountCount > 0 ? insertAccountStatement : null,
+                updateAccountCount > 0 ? updateAccountStatement : null,
+                insertContainerCount > 0 ? insertContainerStatement : null,
+                updateContainerCount > 0 ? updateContainerStatement : null);
+            conn.commit();
+            batchCount = 0;
+            insertAccountCount = 0;
+            updateAccountCount = 0;
+            insertContainerCount = 0;
+            updateContainerCount = 0;
+          }
+        }
+
+        // Commit transaction if one of the following is true. Else, we continue to add more #accounts/containers in current batch.
+        // a) If this is the last account in the list.
+        // b) If subsequent account (and its containers) cannot fit in remaining slots of current batch.
+        if (batchCount > 0 && (i == accountsInfo.size() - 1 ||
+            accountsInfo.get(i + 1).getAddedContainers().size() + accountsInfo.get(i + 1).getModifiedContainers().size()
+                + 1 > (batchSize - batchCount))) {
+          executeBatchUpdates(insertAccountCount > 0 ? insertAccountStatement : null,
+              updateAccountCount > 0 ? updateAccountStatement : null,
+              insertContainerCount > 0 ? insertContainerStatement : null,
+              updateContainerCount > 0 ? updateContainerStatement : null);
+          conn.commit();
+          batchCount = 0;
+          insertAccountCount = 0;
+          updateAccountCount = 0;
+          insertContainerCount = 0;
+          updateContainerCount = 0;
+        }
+      }
+
+      dataAccessor.onSuccess(Write, System.currentTimeMillis() - startTimeMs);
+    } catch (SQLException e) {
+      //rollback
+      conn.rollback();
+      dataAccessor.onException(e, Write);
+      throw e;
+    }
+
+    // reset auto commit to true
+    conn.setAutoCommit(true);
+  }
+
+  /**
+   * Executes batch inserts and updates for input {@link PreparedStatement}s.
+   * @param insertAccountStatement {@link PreparedStatement} for inserting {@link Account}s
+   * @param updateAccountStatement {@link PreparedStatement} for updating {@link Account}s
+   * @param insertContainerStatement {@link PreparedStatement} for inserting {@link Container}s
+   * @param updateContainerStatement {@link PreparedStatement} for updating {@link Container}s
+   * @throws SQLException
+   */
+  public void executeBatchUpdates(PreparedStatement insertAccountStatement, PreparedStatement updateAccountStatement,
```

Comment:
Also, since the method doesn't distinguish between the statements, it could take a variable array like:
private void executeBatchUpdates(PreparedStatement ... statement)